tests, Fix networkpolicy tests and support them upstream

oshoval opened the pull request:
Networkpolicy tests were skipped because when those tests were created, k8s didnt support them.
Now K8s does support them (2 years passed since).

Fixing Flakiness by using waiting for the nettwork policy to finish its async creation (using eventually on the ping).
Use PingFromVMConsole robust method.
Skip Kind providers, since its not supported on kubevirtci kind lanes out of the box.
Wait for Network Policy deletion, in order not to mix policies between tests.
Refactor tests.

Signed-off-by: Or Shoval <oshoval@redhat.com>

```release-note
None
```

oshoval commented:
how should we test it ?
it needs ocp + ovs

phoracek commented:
@oshoval Quique is now working on conversion of these tests to be runnable on Calico. After that, we can move forward with this one.

oshoval commented:
> @oshoval @qinqon is now working on conversion of these tests to be runnable on Calico. After that, we can move forward with this
> one.

ok thanks
adding hold meanwhile i guess
/hold

oshoval commented:
@phoracek 
@EdDev 
i think meanwhile we can do code review on how it should look, even if we dont merge until its tested?

oshoval commented:
/test pull-kubevirt-check-tests-for-flakes

oshoval commented:
/hold cancel
Skip was safe to remove, thanks Quique

oshoval commented on tests/networkpolicy_test.go:
```diff
@@ -71,6 +71,20 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 		vmic, err = virtClient.VirtualMachineInstance(tests.NamespaceTestAlternative).Get(vmic.Name, &v13.GetOptions{})
 		Expect(err).ToNot(HaveOccurred())
 
+		By("Checking that the VirtualMachineInstance vmia console has expected output, and configurating ipv6")
```
can be refactored to use a loop, or other shortcuts.

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -71,6 +71,20 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 		vmic, err = virtClient.VirtualMachineInstance(tests.NamespaceTestAlternative).Get(vmic.Name, &v13.GetOptions{})
 		Expect(err).ToNot(HaveOccurred())
 
+		By("Checking that the VirtualMachineInstance vmia console has expected output, and configurating ipv6")
+		expecter, err := tests.LoggedInCirrosExpecter(vmia)
+		Expect(err).ToNot(HaveOccurred())
+		expecter.Close()
+
+		By("Checking that the VirtualMachineInstance vmib console has expected output, and configurating ipv6")
+		expecter, err = tests.LoggedInCirrosExpecter(vmib)
+		Expect(err).ToNot(HaveOccurred())
+		expecter.Close()
+
+		By("Checking that the VirtualMachineInstance vmic console has expected output, and configurating ipv6")
```
Instead of `WaitForSuccessfulVMIStart` you can use `WaitUntilVMIReady` this takes a expector factory function so it exactly what we need here.

oshoval replied:
yep thx this what i meant with "or other shortcuts"
https://github.com/kubevirt/kubevirt/pull/4045#discussion_r476508946

oshoval replied:
+@EdDev 
+Quique

updating it, doesnt work out of the box, i believe that the issue is fix like #4000 presented

but soon we are going to use cloud-init, is it safe then to not do this refactor atm please ?

qinqon replied:
I am refactoring how we run commands that's does not affect where do we take ips from, right ?

oshoval replied:
> I am refactoring how we run commands that's does not affect where do we take ips from, right ?

True, but wonder if it will solve the flakiness you said you saw, and that i am saw here.
I will first finish the other PR i am on and then we can resync, just we wont do the same (solving flakes in this test family)

EdDev replied:
Sorry, I do not understand what refactoring you talk about. This is a new code... login into the vmi and defining the IPv6. So this is about doing it right, not about refactoring.

I also do not understand how the usage of the wrapper helper does not work and this one does work... they are suppose to do the exact same thing, no?

oshoval replied:
> Sorry, I do not understand what refactoring you talk about. This is a new code... login into the vmi and defining the IPv6. So this is about doing it right, not about refactoring.
> 
> I also do not understand how the usage of the wrapper helper does not work and this one does work... they are suppose to do the exact same thing, no?

It doesnt work out the box, even if "supposed" , i have assumption that adaption as PR 4000 did should solve it.
i think that even if its a new code, the current condition isnt bad, and we are soon moving to cloud-init, (very soon), so i dont see a reason to spend time on it, but first focus on the important things, such are flakiness or other tasks, but whatever you choose.
In my opinion since we know the code is going to be refacotred pretty soon (cloud init usage), i would suggest to consider the current use, as we did not once.

oshoval replied:
found the problem, WaitUntilVMIReady is for default NS, fixing

oshoval replied:
Done, but just this, will take care now of the other commits, dont review please yet.
About flakiness, from what i understood Quique found the problem, so i will wait for him.

maiqueb commented on tests/networkpolicy_test.go:
```diff
@@ -104,14 +94,18 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmib")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingFail(ip, vmib)
+			Expect(vmia.Status.Interfaces[0].IPs).NotTo(BeEmpty())
+			for _, ip := range vmia.Status.Interfaces[0].IPs {
+				assertPingFail(ip, vmib)
+			}
```
Please encapsulate this into a function, then use it.

oshoval replied:
you mean to encap this right ?
```
			for _, ip := range vmia.Status.Interfaces[0].IPs {
				assertPingFail(ip, vmib)
			}
```

oshoval replied:
done, if needed will squash the commit to other one

maiqueb commented on tests/networkpolicy_test.go:
```diff
@@ -147,14 +141,18 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmib in same namespace")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingSucceed(ip, vmib)
+			Expect(vmia.Status.Interfaces[0].IPs).NotTo(BeEmpty())
+			for _, ip := range vmia.Status.Interfaces[0].IPs {
+				assertPingSucceed(ip, vmib)
+			}
```
Please encapsulate this into a function, then use it.

oshoval replied:
done, if needed will squash the commit to other one

maiqueb commented on tests/networkpolicy_test.go:
```diff
@@ -44,7 +44,6 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 		tests.PanicOnError(err)
 
 		tests.SkipIfUseFlannel(virtClient)
-		tests.SkipIfNotUseNetworkPolicy(virtClient)
```
Please mention in this commit why this can be removed. I honestly do not understand why.

oshoval replied:
@qinqon found out it actually works on k8s cluster (but flakey)
the original commit that added the skip 2 years ago along with the tests as they are on master,
said that back then it was supported on openshift only,
so its fine to mention its safe to run it on U/S and the skip isnt needed right ?

oshoval replied:
done

maiqueb replied:
OK, I understand. 

Do you happen to know what changed - i.e. how are we able to run this now on k8s clusters ? If so, add it to the commit msg.

oshoval commented:
Ready to review, PTAL

oshoval commented:
Thanks Quique for the eventually solution

oshoval commented:
/test pull-kubevirt-e2e-k8s-cnao-1.17

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -15,28 +15,23 @@ import (
-	defer expecter.Close()
+func assertPingSucceedBetweenVMs(vmisrc, vmidst *v1.VirtualMachineInstance) {
+	for _, ip := range vmidst.Status.Interfaces[0].IPs {
+		pingEventually(vmisrc, ip).Should(Succeed())
```
Can we add the "should" string at the end of `Succeed()` ?

oshoval replied:
Done

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -15,28 +15,23 @@ import (
-	Expect(err).ToNot(HaveOccurred())
+func assertPingFailBetweenVMs(vmisrc, vmidst *v1.VirtualMachineInstance) {
+	for _, ip := range vmidst.Status.Interfaces[0].IPs {
+		pingEventually(vmisrc, ip).ShouldNot(Succeed())
```
ditto

oshoval replied:
Done

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -15,28 +15,23 @@ import (
-	}, 60)
-	Expect(err).ToNot(HaveOccurred())
+func pingEventually(fromVmi *v1.VirtualMachineInstance, toIp string) AsyncAssertion {
+	return Eventually(func() error {
```
Change to `EventuallyWithOffset(2, ....` so after failing ginkgo will point to the failing tests not the helper function.

oshoval replied:
Done

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -63,25 +57,17 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 
+		vmia = tests.WaitUntilVMIReady(vmia, tests.LoggedInCirrosExpecter)
+		vmib = tests.WaitUntilVMIReady(vmib, tests.LoggedInCirrosExpecter)
+		vmic = tests.WaitUntilVMIReadyNamespace(vmic, vmic.Namespace, tests.LoggedInCirrosExpecter)
```
We can chagne utils.go and at `WaitUntilVMIReady` can use `vmi.Namespace` instead of the default test namespace this way we can remove `WaitUntilVMIReadyNamespace` and use `WaitUntilVMIReady` here.

oshoval replied:
Done

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -104,14 +90,14 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmib from vmia")
-			ip := vmib.Status.Interfaces[0].IP
-			assertPingFail(ip, vmia)
+			Expect(vmib.Status.Interfaces[0].IPs).NotTo(BeEmpty())
```
ditto

oshoval replied:
lets continue at
https://github.com/kubevirt/kubevirt/pull/4045#discussion_r478245984

(same deal will be at the end for all the 7 tests)

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -104,14 +90,14 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmib")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingFail(ip, vmib)
+			Expect(vmia.Status.Interfaces[0].IPs).NotTo(BeEmpty())
```
Move this expectation with `ExpectWithOffset(2` into the `pingEventually`

oshoval replied:
`assertPingFailBetweenVMs` does `for _, ip := range vmidst.Status.Interfaces[0].IPs`
before it call `pingEventually`, so it cant be part of `pingEventually`,

Moroever, i think we even shouldnt move it into assertPingFailBetweenVMs (or its counter part) success
because its a preliminary check that the test is doing.

Same answer for all the tests because they have same comment.

qinqon replied:
Well test is testing ping from vmiab to vmia, so I don't expect to see explicitly a checking on IPs, I would do that as part of `pingEventually` with the benefict of not repeating that at all the test.

oshoval replied:
pingEventually test just one ip at, the for is above it, so at least lets do it one level above?
another thing is that first will work on the change Alon and Miguel asked, and then we even might remove the eventually, and so on from the ping (but can adapt this change in the new structure)

oshoval replied:
Done (for all 7 tests)

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -147,14 +133,14 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmib in same namespace")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingSucceed(ip, vmib)
+			Expect(vmia.Status.Interfaces[0].IPs).NotTo(BeEmpty())
```
ditto

oshoval replied:
https://github.com/kubevirt/kubevirt/pull/4045#discussion_r478245984

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -147,14 +133,14 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmic in differnet namespace")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingFail(ip, vmic)
+			Expect(vmia.Status.Interfaces[0].IPs).NotTo(BeEmpty())
```
ditto

oshoval replied:
https://github.com/kubevirt/kubevirt/pull/4045#discussion_r478245984

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -186,20 +172,20 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmic")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingFail(ip, vmic)
+			Expect(vmia.Status.Interfaces[0].IPs).NotTo(BeEmpty())
```
ditto

oshoval replied:
https://github.com/kubevirt/kubevirt/pull/4045#discussion_r478245984

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -186,20 +172,20 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmib")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingFail(ip, vmib)
+			Expect(vmia.Status.Interfaces[0].IPs).NotTo(BeEmpty())
```
ditto

oshoval replied:
https://github.com/kubevirt/kubevirt/pull/4045#discussion_r478245984

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -186,20 +172,20 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmib from vmic")
-			ip := vmib.Status.Interfaces[0].IP
-			assertPingSucceed(ip, vmic)
+			Expect(vmib.Status.Interfaces[0].IPs).NotTo(BeEmpty())
```
ditto

oshoval replied:
https://github.com/kubevirt/kubevirt/pull/4045#discussion_r478245984

maiqueb commented on tests/networkpolicy_test.go:
```diff
@@ -52,25 +52,17 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
+		vmic, err = virtClient.VirtualMachineInstance(tests.NamespaceTestAlternative).Create(vmic)
 		Expect(err).ToNot(HaveOccurred())
 
+		vmia = tests.WaitUntilVMIReady(vmia, tests.LoggedInCirrosExpecter)
+		vmib = tests.WaitUntilVMIReady(vmib, tests.LoggedInCirrosExpecter)
+		vmic = tests.WaitUntilVMIReadyNamespace(vmic, vmic.Namespace, tests.LoggedInCirrosExpecter)
```
I think it's preferable to just create an helper to do this, something like what @phoracek did in https://github.com/kubevirt/kubevirt/blob/master/tests/network/primary_pod_network.go#L102 

I'm also copying that 'pattern' in other tests - see [this discussion](https://github.com/kubevirt/kubevirt/pull/4035#discussion_r478201237) about keeping the helper names coherent.

oshoval replied:
done

maiqueb commented on tests/networkpolicy_test.go:
```diff
@@ -12,35 +15,25 @@ import (
 	cd "kubevirt.io/kubevirt/tests/containerdisk"
 )
 
+func pingEventually(fromVmi *v1.VirtualMachineInstance, toIp string) AsyncAssertion {
```
This will potentially increase - by a lot - the execution time of failed builds ...

I don't like it.

oshoval replied:
see please https://github.com/kubevirt/kubevirt/pull/4045#issuecomment-681897629

oshoval replied:
doesnt exists anymore, due to what @alonSadan suggested (and you +1), no need for it anymore.

maiqueb reviewed (commented):
I'm concerned about the ping eventually approach.

Why was that not required until now ?... This has the potential to bloat - by a lot - the execution times on failed builds.

oshoval commented:
> I'm concerned about the ping eventually approach.
> 
> Why was that not required until now ?... This has the potential to bloat - by a lot - the execution times on failed builds.

Those tests are were on skip this is the reason..
it takes a bit time for the policy to be applied, it uses Edy's mechanism that already saves time even when ping failed, because it wont wait for the timeout.

There is one bug anyway that i do need to fix beside the comments.

alonSadan commented:
> > I'm concerned about the ping eventually approach.
> > Why was that not required until now ?... This has the potential to bloat - by a lot - the execution times on failed builds.
> 
> Those tests are were on skip this is the reason..
> it takes a bit time for the policy to be applied, it uses Edy's mechanism that already saves time even when ping failed, because it wont wait for the timeout, but even stop on retcode != 0, instead of timeout.

is it possiable that the  Eventuallys will not be on the ping but on the appliance of the policy, because the way I see it, those are two different checks. I think that Eventuallys scope should be limited as possible because the tend to cover for bugs and here maybe(not sure) we should limit it to the policy appliance. is there a way to check that the policy was applied without pinging, and to use the Eventually just for the policy appliance check? like try to GET it, or something else, or ping is the right way?
> 
> There is one bug anyway that i do need to fix beside the comments.

oshoval commented on tests/networkpolicy_test.go:
```diff
@@ -20,23 +17,16 @@ func assertPingSucceed(ip string, vmi *v1.VirtualMachineInstance) {
-		&expect.BExp{R: "0% packet loss"},
-	}, 60)
-	Expect(err).ToNot(HaveOccurred())
+	Expect(tests.PingFromVMConsole(vmi, ip, "-w 3")).To(Succeed())
```
ping module provides the `PingFromVMConsole`
this 2 functions you said now are changed, and there are 2 new functions that just this file use.
(we can as part of follow PR if possible please, to look on higher view and encapsulate couple similar usages, but Quique already started some refactoring of ping module i think, so i would ask please to wait he finish if possible)

(i might refactor the commits, it looks like you wrote on an outdated one)

oshoval commented:
> > > I'm concerned about the ping eventually approach.
> > > Why was that not required until now ?... This has the potential to bloat - by a lot - the execution times on failed builds.
> > 
> > 
> > Those tests are were on skip this is the reason..
> > it takes a bit time for the policy to be applied, it uses Edy's mechanism that already saves time even when ping failed, because it wont wait for the timeout, but even stop on retcode != 0, instead of timeout.
> 
> Is it possible that the Eventuallys will not be on the ping but on the appliance of the policy, because the way I see it, those are two different checks. I think that Eventuallys scope should be limited as possible because the tend to cover for bugs and here maybe(not sure) we should limit it to the policy appliance. is there a way to check that the policy was applied without pinging, and to use the Eventually just for the policy appliance check? like try to GET it, or something else, or ping is the right way?
> 
> > There is one bug anyway that i do need to fix beside the comments.

Its possible, i want please first to fix the little bug that there is, and the comments,
and if possible please to consider it in a follow PR, because in case the policy wont be applied, we would catch that the ping wont success.

maiqueb commented:
> > > I'm concerned about the ping eventually approach.
> > > Why was that not required until now ?... This has the potential to bloat - by a lot - the execution times on failed builds.
> > 
> > 
> > Those tests are were on skip this is the reason..
> > it takes a bit time for the policy to be applied, it uses Edy's mechanism that already saves time even when ping failed, because it wont wait for the timeout, but even stop on retcode != 0, instead of timeout.
> 
> Is it possible that the Eventuallys will not be on the ping but on the appliance of the policy, because the way I see it, those are two different checks. I think that Eventuallys scope should be limited as possible because the tend to cover for bugs and here maybe(not sure) we should limit it to the policy appliance. is there a way to check that the policy was applied without pinging, and to use the Eventually just for the policy appliance check? like try to GET it, or something else, or ping is the right way?

+1; I think you should get the VM "ready" (you an have your eventually there), then do a single ping. That's the pattern we have been using, and expresses better what we want to achieve.

> 
> > There is one bug anyway that i do need to fix beside the comments.

alonSadan commented on tests/networkpolicy_test.go:
```diff
@@ -20,23 +17,16 @@ func assertPingSucceed(ip string, vmi *v1.VirtualMachineInstance) {
-		&expect.BExp{R: "100% packet loss"},
-	}, 60)
-	Expect(err).ToNot(HaveOccurred())
+	err = tests.PingFromVMConsole(vmi, ip, "-w 3")
```
nit: the `Expect(Something).To(Succeed())` pattern can be used here to: 

`Expect(tests.PingFromVMConsole(vmi, ip, "-w 3")).To(Succeed())`

oshoval replied:
this is outdated, function was removed and changed

oshoval commented:
just rebased upon Quique request

oshoval commented:
just rebased upon Quique request

oshoval commented:
just rebased upon Quique request

oshoval commented:
Did part of the comments, still need to do some, marked the done ones inline their thread.

oshoval commented:
Done all, will rearrange commits once we i get your blessing.
Thanks

oshoval commented:
/test pull-kubevirt-check-tests-for-flakes

oshoval commented:
/test pull-kubevirt-check-tests-for-flakes
Passed when there is both eventually on the ping and on the GET of the policy,
retesting to see if it passes couple times

oshoval commented:
/test pull-kubevirt-check-tests-for-flakes
(passed twice, want to see more times that its stable)

oshoval commented:
All the tests that check ping fail are failing on Kind IPv6 (1.18 locally passes),
Can we add a skip for kind ?
I see this on the docs of Ingress field
```
	// List of ingress rules to be applied to the selected pods. Traffic is allowed to
	// a pod if there are no NetworkPolicies selecting the pod
	// (and cluster policy otherwise allows the traffic), OR if the traffic source is
	// the pod's local node, OR if the traffic matches at least one ingress rule
	// across all of the NetworkPolicy objects whose podSelector matches the pod. If
	// this field is empty then this NetworkPolicy does not allow any traffic (and serves
	// solely to ensure that the pods it selects are isolated by default)
	// +optional
```
Maybe the "OR if the traffic source is the pod's local node" is the reason.
Will try kind locally to see if this is the same behavior all the time.

RamLavi commented on tests/networkpolicy_test.go:
```diff
@@ -52,25 +52,17 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 
 		vmib = tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n")
-		_, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmib)
+		vmib, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(vmib)
```
out of curiosity, why is this change needed? does Create add something that is needed?

EdDev commented:
The pattern of adding review changes as individual commits at the tail is not something I can work with when reviewing. I am passing one commit at a time in order to limit my focus to a specific change and not to everything as a whole, I cannot practice that with this approach.

If this will change in this PR, let me know and I'll have a look.

RamLavi commented on tests/networkpolicy_test.go:
```diff
@@ -12,35 +15,25 @@ import (
+func pingEventually(fromVmi *v1.VirtualMachineInstance, toIp string) AsyncAssertion {
+	return Eventually(func() error {
+		By(fmt.Sprintf("Pinging from VMI %s/%s to Ip %s", fromVmi.Namespace, fromVmi.Name, toIp))
+		return tests.PingFromVMConsole(fromVmi, toIp)
```
yes as long as we don't change behavior. ping should have a timeout.

oshoval replied:
we wont change it so fast now that we have the new mechanism, there is timeout in the mechanism itself, all the code use this, those arguments were there when it didnt use the mechanism

EdDev replied:
The helper has a default timeout of 5sec (-w) and an expected timeout of 10sec.
By removing the `-w 3` you added 2sec to this test in case it fails.

If it is a matter of retry you can just drop the `-c` and there is no need for the eventually.

oshoval replied:
as we spoke we need the eventually at least on negative tests, because the first ping might success, until the policy is effective and blocking the traffic.
a successful ping would return immediately, and wont consume even 3 or 5 seconds.

on the positive tests, i will check what can be done, in case its important.

oshoval commented:
> The pattern of adding review changes as individual commits at the tail is not something I can work with when reviewing. I am passing one commit at a time in order to limit my focus to a specific change and not to everything as a whole, I cannot practice that with this approach.
> 
> If this will change in this PR, let me know and I'll have a look.

From what i understood from our chat now and back then, middle rearrangements wasn't the preferred method by couple Authors, and depends on the reviewer.
I will just rearrange it now as it should be in the end, since it would be easier for me, and needs to be done anyhow (thought just to do it in the end, but upon your request will do it before).

oshoval commented:
Rearranged the commits, PTAL
Removed the GET, as we spoke in the meeting, it doesn't fail, and just added delay to the Ping
so extended the ping eventually timeout.
Added skip on kind.
Will run several times flake finder on the PR to see its stable as did before.

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -15,28 +15,43 @@ import (
+}
+
+func waitForNetworkPolicyDeletion(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, policyName string) {
+	Eventually(func() error {
+		_, err := virtClient.NetworkingV1().NetworkPolicies(vmi.Namespace).Get(policyName, v13.GetOptions{})
+		return err
```
We need to check the `IsNotFound` error to be correct here similar to https://github.com/nmstate/kubernetes-nmstate/blob/master/test/e2e/handler/utils.go#L175-L178

oshoval replied:
Nice, thanks, done (both comments) PTAL

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -15,28 +15,43 @@ import (
+	Eventually(func() error {
+		_, err := virtClient.NetworkingV1().NetworkPolicies(vmi.Namespace).Get(policyName, v13.GetOptions{})
+		return err
+	}, 10*time.Second).ShouldNot(Succeed())
```
Set interval to 1 second with `}, 10*time.Second, time.Second)` by default is too small

qinqon reviewed (changes requested):
Looks like some comments on checking delete from what we have learn at other projects.

oshoval commented:
/test pull-kubevirt-check-tests-for-flakes 
passed 1/1 times so far

oshoval commented:
/test pull-kubevirt-check-tests-for-flakes 
passed 2/2

qinqon commented:
/retest

oshoval commented:
Fixed according Quique comments, (waitForNetworkPolicyDeletion func).
Rebase.
PTAL

qinqon commented on tests/networkpolicy_test.go:
```diff
@@ -2,41 +2,58 @@ package tests_test
+	return tests.WaitUntilVMIReady(vmi, tests.LoggedInCirrosExpecter)
+}
+
+func waitForNetworkPolicyDeletion(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, policyName string) {
+	Eventually(func() bool {
+		_, err := virtClient.NetworkingV1().NetworkPolicies(vmi.Namespace).Get(policyName, v13.GetOptions{})
+		return apierrors.IsNotFound(err)
+	}, 10*time.Second, time.Second).Should(BeTrue())
```
Let's make it more verbose to make easier to debug, my previous reference was not totally right the more verbose version of this would be https://github.com/nmstate/kubernetes-nmstate/blob/master/test/e2e/handler/utils.go#L186-L194

verbose means, printing the not matched error to know what `.Get` is returning and adding an description string at `.Should`

oshoval replied:
Done thx

qinqon commented:
/lgtm

EdDev commented on tests/utils.go:
```diff
@@ -2669,7 +2669,7 @@ func WaitUntilVMIReady(vmi *v1.VirtualMachineInstance, expecterFactory VMIExpect
 	// Fetch the new VirtualMachineInstance with updated status
 	virtClient, err := kubecli.GetKubevirtClient()
-	vmi, err = virtClient.VirtualMachineInstance(NamespaceTestDefault).Get(vmi.Name, &metav1.GetOptions{})
+	vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
```
If you decided to fix one, then you should fix all (for all Get and Delete). Having only one fixed make no sense.
(in this file)

If you want to do it, I suggest you do it on a separate PR to avoid other surprises.

oshoval replied:
i have to fix this one, because i use WaitUntilVMIReady for non default namespace.
I dont think we need to fix every occurrence on this PR ?
and even if so, lets keep this one in this PR if possible please and fix all the others on other PR?
(else i must first create a PR fixing all, and then use it in this PR)

EdDev replied:
Then squash it as part of fixing the existing tests... If you take this to its own commit, from mys point of view you should fix them all.

oshoval replied:
ok will then squash it (or put as part of the fixing tests), and split the last commit in other PR as you suggested, thanks

oshoval replied:
i will then squash it as part of the refactor in the dual stack separate PR, based on the last commit and this commit,
since i needed it because the refactor.

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -39,6 +41,16 @@ func assertPingFail(ip string, vmi *v1.VirtualMachineInstance) {
 	Expect(err).ToNot(HaveOccurred())
 }
 
+func waitForNetworkPolicyDeletion(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, policyName string) {
```
Please place it at the tail. It is easier to see the caller before the callee when reading the code.

There is also no need to pass the virtClient, just fetch it here (e.g. `virtCli, err := kubecli.GetKubevirtClient()`)

oshoval replied:
placed up , just because the rest of the funcs are there (usually i put it in the tail),
and about the virtClient, the caller already has it, so saw that this is what other functions do in case it already exists

but will do those 2 changes.

RamLavi commented on tests/networkpolicy_test.go:
```diff
@@ -54,6 +54,7 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 		tests.PanicOnError(err)
 
```
This commit description line is not clear to me: We first focus on dual stack and single stack lanes.
Do you mean to say that this commit only handles skipping NetworkPolicy tests on dual stack and single stack lanes?

oshoval replied:
it only skip on `Kind`
single and dual are not skipped.

RamLavi replied:
ok but the wording "We first focus on dual stack and single stack lanes" is not very clear. 
could you reword this sentence?

RamLavi commented on tests/networkpolicy_test.go:
```diff
@@ -9,6 +9,8 @@ import (
 	v1network "k8s.io/api/networking/v1"
 	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
```
this import  should be joined to the prior group, and usually ginkgo groups are also grouped alone:

```
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"

	expect "github.com/google/goexpect"
	v1network "k8s.io/api/networking/v1"
	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
```

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -9,6 +9,8 @@ import (
 	v1network "k8s.io/api/networking/v1"
 	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
```
I saw no others using this alias, they just use it without and mentioning plain `errors`.

oshoval replied:
nmstate use it, but i will see what you suggested and what kubevirt use

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -39,6 +41,16 @@ func assertPingFail(ip string, vmi *v1.VirtualMachineInstance) {
+			Fail(fmt.Sprintf("Unexpected error waiting for policy deletion: %v", err))
+		}
+		return apierrors.IsNotFound(err)
+	}, 10*time.Second, time.Second).Should(BeTrue(), fmt.Sprintf("Policy %s was not deleted", policyName))
```
Please see [here](https://github.com/kubevirt/kubevirt/blob/master/tests/network/primary_pod_network.go#L123) on how to assert this and avoid this extra Should text.

qinqon replied:
Interesting...

oshoval replied:
```
		if err != nil && !errors.IsNotFound(err) {
			Fail(fmt.Sprintf("Unexpected error waiting for policy deletion: %v", err))
		}
```
Removing this then as well

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -39,6 +41,16 @@ func assertPingFail(ip string, vmi *v1.VirtualMachineInstance) {
 }
 
+func waitForNetworkPolicyDeletion(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, policyName string) {
+	Eventually(func() bool {
```
If you use the assert down here, use an offset please.

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -116,15 +129,18 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 
 		AfterEach(func() {
 			Expect(virtClient.NetworkingV1().NetworkPolicies(vmia.Namespace).Delete("deny-by-default", &v13.DeleteOptions{})).To(Succeed())
```
This also needs fixing, it is incorrect to delete something that was not yet created (e.g. if something failed before a creation occurred). True for all the rest as well.

EdDev reviewed (changes requested):
I reviewed all commits except the last one.
IMO the last one merits a different PR as all the rest are fixing the existing tests while the last is adding dual stack support.

qinqon commented:
/retest

RamLavi commented on tests/networkpolicy_test.go:
```diff
@@ -9,6 +9,8 @@ import (
 	v1network "k8s.io/api/networking/v1"
```
intended for commit `tests, Add waitForNetworkPolicyDeletion` :

Does this commit intend to fix flakiness? 
If so, please state it on the commit desc.
if not, then I'm not sure why it is relevant to "making networkpolicy dual-stack compatible".
can you please elaborate?

oshoval replied:
it can fix flakiness (not the only one that was here) because the nature of the policies, 
each group of test there use a common policy and expect only this policy to exists, we dont want leftovers from previous tests that might cause an unexpected behavior.
each `Context` is based only on the policy its create and assumes only this policy exists.

As part of the effort, those fixes were applied, i might move the last commit to other PR as Edy suggested, and it would be less confusing.
its not about dual stack, it was a fix to make the tests better and more stable.

RamLavi replied:
ok but if you keep this commit in the PR then please mention in the desc. that this commit means to fix flakiness. 
will make it easier for reviewer to have the proper context

oshoval replied:
Done

oshoval commented:
Did all comments of Ram and Edy,
removed the last commit (dual stack + refactor for it) and the WaitUntilVMIReady commit, will create a new PR for those.
Testing locally, will squash the new commit once there is a blessing.
PTAL

oshoval commented:
Did all comments of Ram and Edy,
removed the last commit (dual stack + refactor for it) and the WaitUntilVMIReady commit, will create a new PR for those.
Testing locally, will squash the new commit once there is a blessing.

But
I have a mistake, i need to add here the pingEventually stuff, dont review yet please

oshoval commented:
Added last commit that fix flakes, PTAL
one commit need to be squashed to the one below it (there i answered the latest requests)

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -111,14 +65,12 @@ var _ = Describe("[rfe_id:150][crit:high][vendor:cnv-qe@redhat.com][level:compon
 			By("Connect vmia from vmib")
-			ip := vmia.Status.Interfaces[0].IP
-			assertPingFail(ip, vmib)
+			pingEventually(vmib, vmia.Status.Interfaces[0].IP).ShouldNot(Succeed())
```
I think this will not provide the expected result.
Consider a case where the ping fist succeeds (because the policy is not yet deployed) and only after 5sec fails.
In such a case, the eventually will immediately pass and we will fail the assert (when we should have waited instead).

oshoval replied:
the default parameters are -c 1 -w 5 as you know, 
each success will return but as its wrapped in eventually it will retry each 1 seconds (as its the interval of Eventually),
so there are around 15 retries (because the total retry is 15 seconds), 
i saw by practice that 15 seconds is enough and there are no falkes.

Also talked with the community and there isnt a way to validate ingress easily, 
the CNI watch for pods and creates the rules in an async matter

EdDev replied:
Now I see that the helper does not assert, just returns the result and the assertion here will cause a retry. So the scenario I used above will work fine.

Still I am worry here with the introduction of this retry.
Lets check the "Should(Succeed())" scenario: What if it succeeds at the beginning and then blocks when the policy is deployed? We will never detect this in case we actually expect it to pass in the end.
It can also work the other way around: We expect it to fail in general, but it fails only at the begining and when the policy is applied it passes, this is another case where we will miss the problem.

Are there such a scenarios here?

I guess that compared to the current asserts, this retry is not worse, the question is if it is good enough.

oshoval replied:
lets split it to the expect ping fail and expect ping success, for simplicity.

expect pass:
( the scenario you said),
we can increase the "-c 12 -w 3", this way it would check ping for longer time

expect fail:
a policy should stop the traffic,
we are pinging until the first stop, we can add a sanity that 3 stops should occur, i dont think it must be at this very moment imho, i feel what we did is good enough for now, we dont check sanity of checks on at least few other places as well.

oshoval commented:
Hopefully answered all Edy latest comments.
Arranged commits.
Rebased.
PTAL.

qinqon commented:
/lgtm

qinqon commented on tests/utils.go:
```diff
@@ -4852,6 +4843,12 @@ func IsRunningOnKindInfraIPv6() bool {
 	return strings.HasPrefix(provider, "kind-k8s-1.17.0-ipv6")
 }
 
+func SkipNetworkPolicyRunningOnKindInfra() {
```
Sorry I miss this, we use this only on networkpolicy_test let's move it there, so we don't increment utils.go

qinqon reviewed (changes requested):
Let's not increment utils.go

oshoval commented:
Updated according Quique and Ram comments.

qinqon commented:
/lgtm

RamLavi commented:
/lgtm

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -215,3 +225,18 @@ func skipNetworkPolicyRunningOnKindInfra() {
+	}
+
+	virtClient, err := kubecli.GetKubevirtClient()
+	Expect(err).ToNot(HaveOccurred())
```
nit: You forgot an offset here. (if you change something else, you can change this one as well, otherwise do not bother)

EdDev commented on tests/networkpolicy_test.go:
```diff
@@ -226,6 +176,31 @@ func skipNetworkPolicyRunningOnKindInfra() {
+	return vmi
+}
+
+func checkPingSuccess(fromVmi *v1.VirtualMachineInstance, toIp string) {
```
I think the original name is more correct: `assertPingSucceed` as it is indeed asserting.
But I would argue that we do not need a helper in this case at all, avoiding the need to add an offset.

I also have a behavior question here: What happens if the first two attempts fail and the rest succeed, will that end with a return value of "0"/success?

oshoval replied:
> I think the original name is more correct: `assertPingSucceed` as it is indeed asserting.
> But I would argue that we do not need a helper in this case at all, avoiding the need to add an offset.
> 

Thought about it as well, but since it repeats in two times, and because the next pending PR #4099 
will add logic, and to have a symmetric API between ping fail and ping success, decided to put a function
(i understand one can argue that each PR on its on, see please the other reasons).
Will change the name if needed either on this PR or in the follow one #4099 up to you.
 
> I also have a behavior question here: What happens if the first two attempts fail and the rest succeed, will that end with a return value of "0"/success?

no, because we dont have eventually and ping "-w 8" pings 8 seconds deadline and return success iff (not typo this time)
all successed.

qinqon replied:
Stupid question, what about implementing it with Eventually Consistently so we don't depend on ping behaviour ? it will make it consistently with `checkPingFailure`

oshoval replied:
already done, PTAL
fixed both threads

EdDev replied:
I see this from the man:

> If  ping does not receive any reply packets at all it will exit with code 1. If a packet count and deadline are both specified, and fewer than count packets are received by the time the deadline has
> arrived, it will also exit with code 1.  On other error it exits with code 2. Otherwise it exits with code 0. This makes it possible to use the exit code to see if a host is alive or not.
> 

So if -c and -w are set, it should be the case. Can you test it?

oshoval replied:
> I see this from the man:

i can but as Quique and me suggested, and since PingFromVMConsole is limited to 10 seconds atm,
and we need more than 10 seconds (because in order to see failures we saw 10 seconds were still flaky), 
i think we should use consistently (pushed the change if you can take a look),
but meanwhile trying manually what you said (i believe it works, so let me know if you prefer it over consistently, imo consistently is better)

oshoval replied:
> I see this from the man:
> 
> > If  ping does not receive any reply packets at all it will exit with code 1. If a packet count and deadline are both specified, and fewer than count packets are received by the time the deadline has
> > arrived, it will also exit with code 1.  On other error it exits with code 2. Otherwise it exits with code 0. This makes it possible to use the exit code to see if a host is alive or not.
> 
> So if -c and -w are set, it should be the case. Can you test it?

tested, return 1 if both -w and -c are supplied, 0 if just -w is supplied and some packets success.

but as Quique and me suggested, and since PingFromVMConsole is limited to 10 seconds atm,
and we need more than 10 seconds (because in order to see failures we saw 10 seconds were still flaky),
i think we should use consistently (pushed the change if you can take a look).

wdyt ?

EdDev replied:
1. `Failed` is a verb in the past tense (or an adjective which is not a fit here), I think you should use as the original `Fail`.
2. I do not mind if you use ping application or repeating the console command (although the later is slower), but I want to point out that expecting for the ping to always succeed from start does not cover all scenarios. it depends on the specific case, which in this case may be fine (you should check). I mean, if the policy was in advance permissive, ping should continue working as you just asserted, on the other hand, if the policy was not permissive, and it just "opened" it up, then it will first fails and later pass.

oshoval replied:
fixed point (1) PTAL

EdDev reviewed (approved):
Thanks.

RamLavi commented:
/lgtm

qinqon commented:
/lgtm

AlonaKaplan commented:
Is there a plan to make this test dual stack?

EdDev commented:
> Is there a plan to make this test dual stack?

I guess you mean this one: https://github.com/kubevirt/kubevirt/pull/4099

oshoval commented:
failures are not due to this PR
/retest

AlonaKaplan commented:
/approve

oshoval commented:
/retest

oshoval commented:
/retest

oshoval commented:
/retest
